diff -Naurp binutils/size.c binutils/size.c
--- binutils/size.c	2012-02-27 11:45:17.000000000 +0530
+++ binutils/size.c	2012-02-27 11:49:07.000000000 +0530
@@ -91,6 +91,7 @@ static int return_code = 0;
 #define AVR128 128UL
 #define AVR256 256UL
 #define AVR512 512UL
+#define AVR2112 2112UL
 #define AVR1K 1024UL
 #define AVR2K 2048UL
 #define AVR4K 4096UL
@@ -193,6 +194,9 @@ avr_device_t avr[] =
 	{"at90scr100",    AVR64K,  AVR4K,  AVR2K},
 	{"at90usb646",    AVR64K,  AVR4K,  AVR2K},
 	{"at90usb647",    AVR64K,  AVR4K,  AVR2K},
+	{"ata5790",       AVR16K,  AVR512, AVR2112},
+	{"ata5790n",      AVR16K,  AVR512, AVR2112},
+	{"ata5795",       AVR8K,   AVR512, AVR2112},
 	{"atmega64",      AVR64K,  AVR4K,  AVR2K},
 	{"atmega64a",     AVR64K,  AVR4K,  AVR2K},
 	{"atmega640",     AVR64K,  AVR8K,  AVR4K},
diff -Naurp gas/config/tc-avr.c gas/config/tc-avr.c
--- gas/config/tc-avr.c	2012-02-27 11:45:17.000000000 +0530
+++ gas/config/tc-avr.c	2012-02-27 11:46:30.000000000 +0530
@@ -168,6 +168,9 @@ static struct mcu_type_s mcu_types[] =
   {"at90pwm3b",  AVR_ISA_AVR4,    bfd_mach_avr4},
   {"at90pwm81",  AVR_ISA_AVR4,    bfd_mach_avr4},
   {"at90pwm161", AVR_ISA_AVR5,    bfd_mach_avr5},
+  {"ata5790",    AVR_ISA_AVR5,    bfd_mach_avr5},
+  {"ata5790n",   AVR_ISA_AVR5,    bfd_mach_avr5},
+  {"ata5795",    AVR_ISA_AVR5,    bfd_mach_avr5},
   {"atmega16",   AVR_ISA_AVR5,    bfd_mach_avr5},
   {"atmega16a",  AVR_ISA_AVR5,    bfd_mach_avr5},
   {"atmega161",  AVR_ISA_M161,    bfd_mach_avr5},
diff -Naurp gas/doc/c-avr.texi gas/doc/c-avr.texi
--- gas/doc/c-avr.texi	2012-02-27 11:45:17.000000000 +0530
+++ gas/doc/c-avr.texi	2012-02-27 11:46:30.000000000 +0530
@@ -77,7 +77,8 @@ atmega649a, atmega649p, atmega6490, atme
 atmega16hvb, atmega16hvbrevb, atmega26hvg, atmega32hvb, atmega32hvbrevb, atmega48hvf, atmega64hve,
 at90can32, at90can64, at90pwm216, at90pwm316, atmega32c1,
 atmega64c1, atmega16m1, atmega32m1, atmega64m1, atmega16u4, atmega32u4,
-atmega32u6, at90usb646, at90usb647, at94k, at90scr100).
+atmega32u6, at90usb646, at90usb647, at94k, at90scr100,
+ata5790, ata5790n, ata5795).
 
 Instruction set avr51 is for the enhanced AVR core with exactly 128K program
 memory space (MCU types: atmega128, atmega128a, atmega1280, atmega1281, atmega1284, atmega1284p,
